src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinProvider.php

```diff
@@ -0,0 +1,425 @@
+    {
+        $filesMap = $this->getFilesMap();
+
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                if (0 === \count($catalogue->all($domain))) {
+                    continue;
+                }
+
+                $content = $this->xliffFileDumper->formatCatalogue($catalogue, $domain, ['default_locale' => $this->defaultLocale]);
+
+                $fileId = $this->getFileIdByDomain($filesMap, $domain);
+
+                if ($catalogue->getLocale() === $this->defaultLocale) {
+                    if (!$fileId) {
+                        $file = $this->addFile($domain, $content);
+                    } else {
+                        $file = $this->updateFile($fileId, $domain, $content);
+                    }
+
+                    $filesMap[$file['name']] = $file['id'];
+                } else {
+                    if (!$fileId) {
+                        continue;
+                    }
+
+                    $this->uploadTranslations($fileId, $domain, $content, $catalogue->getLocale());
+                }
+            }
+        }
+    }
+
+    public function read(array $domains, array $locales): TranslatorBag
+    {
+        $filesMap = $this->getFilesMap();
+
+        $translatorBag = new TranslatorBag();
+
+        foreach ($domains as $domain) {
+            $fileId = $this->getFileIdByDomain($filesMap, $domain);
+
+            if (!$fileId) {
+                continue;
+            }
+
+            foreach ($locales as $locale) {
+                if ($locale !== $this->defaultLocale) {
+                    $content = $this->exportProjectTranslations($locale, $fileId);
+                } else {
+                    $content = $this->downloadSourceFile($fileId);
+                }
+
+                $translatorBag->addCatalogue($this->loader->load($content, $locale, $domain));
+            }
+        }
+
+        return $translatorBag;
+    }
+
+    public function delete(TranslatorBagInterface $translatorBag): void
+    {
+        $filesMap = $this->getFilesMap();
+
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            if ($catalogue->getLocale() !== $this->defaultLocale) {
+                continue;
+            }
+
+            foreach ($catalogue->getDomains() as $domain) {
+                $fileId = $this->getFileIdByDomain($filesMap, $domain);
+
+                if (!$fileId) {
+                    continue;
+                }
+
+                $stringsMap = $this->mapStrings($fileId);
+
+                foreach ($catalogue->all() as $messages) {
+                    foreach ($messages as $id => $message) {
+                        if (!isset($stringsMap[$id])) {
+                            continue;
+                        }
+
+                        $this->deleteString($stringsMap[$id]);
+                    }
+                }
+            }
+        }
+    }
+
+    private function getFilesMap(): array
```

Comment:
I think `getFilesMap` doesn't bring any advantage, it should be merged in `getFileList`, and this one should be called directly.